File: Archived2/Java/Silver/24479.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

public class Main {

    private static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    private static ArrayList<ArrayList<Integer>> adj;
    private static int order = 1;

    private static int startVertex;

    private static int[] visited;

    private static StringBuilder sb = new StringBuilder();

    public static void main(String[] args) throws IOException {

        settingInput();
        dfs(startVertex);
        for (int i=1;i<visited.length;i++) {
            sb.append(visited[i]).append("\n");
        }
        System.out.println(sb);
        br.close();
    }

    private static void settingInput() throws IOException {

        int[] initData = Arrays.stream(br.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();
        adj = new ArrayList<>(initData[0]+1);
        visited = new int[initData[0]+1];
        startVertex = initData[2];
        for (int i=0;i<initData[0]+1;i++) adj.add(new ArrayList<>());
        ArrayList<int[]> edges = new ArrayList<>();
        for (int i=0;i<initData[1];i++) {
            int[] edge = Arrays.stream(br.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();
            edges.add(edge);
            edges.add(new int[] {edge[1], edge[0]});
        }

        edges.sort( (a,b) -> {
            if(a[0] == b[0]) return a[1] - b[1];
            else return a[0]-b[0];
        });

        edges.forEach((edge) -> {
            adj.get(edge[0]).add(edge[1]);
            adj.get(edge[1]).add(edge[0]);
        });

    }

    private static void dfs(int vertex) {
        visited[vertex] = order++;
        adj.get(vertex).forEach((next) -> {
            if(visited[next] == 0) dfs(next);
        });
    }



}

File: Archived2/Kotlin/Silver/24479.kt
import java.io.BufferedReader
import java.io.InputStreamReader


val br = BufferedReader(InputStreamReader(System.`in`))
lateinit var adj: ArrayList<ArrayList<Int>>
var startVertex = 0
var order = 1
lateinit var visited: IntArray

fun main() {

    settingInput()

    dfs(startVertex)

    val result = buildString {
        for(i in 1 until visited.size) {
            append(visited[i]).append("\n")
        }
    }
    println(result)
}

fun settingInput() {

    val initData = br.readLine().split(" ").map { it.toInt() }
    adj = ArrayList()
    startVertex = initData[2]
    visited = IntArray(initData[0]+1)
    repeat(initData[0]+1) {
        adj.add(ArrayList())
    }
    val edges = ArrayList<IntArray>()
    repeat(initData[1]) {
        val edge = br.readLine().split(" ").map { it.toInt() }.toIntArray()
        edges.add(edge)
        edges.add(intArrayOf(edge[1], edge[0]))
    }

    edges.sortWith( compareBy({it[0]}, {it[1]}) )
    edges.forEach {
        adj.get(it[0]).add(it[1])
        adj.get(it[1]).add(it[0])
    }

}

fun dfs(vertex: Int) {
    visited[vertex] = order++
    adj.get(vertex).forEach {
        if (visited[it] == 0) dfs(it)
    }
}